src/model/channel/attachment_type.rs

```diff
@@ -19,75 +19,95 @@ use crate::error::{Error, Result};
+            } => (data, Some(filename)),
             Self::File {
-                file, ..
+                file,
+                filename,
             } => {
-                let mut buf = Vec::new();
-                file.try_clone().await?.read_to_end(&mut buf).await?;
-                buf
+                let data = data_file(file).await?;
+                (data.into(), Some(filename))
             },
             Self::Path(path) => {
-                let mut file = File::open(path).await?;
-                let mut buf = Vec::new();
-                file.read_to_end(&mut buf).await?;
-                buf
+                let filename =
+                    path.file_name().map(|filename| filename.to_string_lossy().to_string());
+                let data = data_path(path).await?;
+
+                (data.into(), filename.map(Cow::from))
             },
             Self::Image(url) => {
-                let response = client.get(url.clone()).send().await?;
-                response.bytes().await?.to_vec()
+                let filename = match url.path_segments().and_then(Iterator::last) {
+                    Some(filename) => filename.to_string().into(),
+                    None => return Err(Error::Url(url.to_string())),
+                };
+
+                let data = data_image(client, url).await?;
+                (data.into(), Some(filename))
             },
-        };
-        Ok(data)
+        })
     }
 
-    pub(crate) fn filename(&self) -> Result<Option<String>> {
-        match self {
+    pub(crate) async fn data(self, client: &Client) -> Result<Cow<'static, [u8]>> {
+        Ok(match self {
             Self::Bytes {
-                filename, ..
-            }
-            | Self::File {
-                filename, ..
-            } => Ok(Some(filename.to_string())),
-            Self::Path(path) => {
-                Ok(path.file_name().map(|filename| filename.to_string_lossy().to_string()))
-            },
-            Self::Image(url) => match url.path_segments().and_then(Iterator::last) {
-                Some(filename) => Ok(Some(filename.to_string())),
-                None => Err(Error::Url(url.to_string())),
-            },
-        }
+                data, ..
+            } => data,
+            Self::File {
+                file, ..
+            } => data_file(file).await?.into(),
+
+            Self::Path(path) => data_path(path).await?.into(),
+            Self::Image(url) => data_image(client, url).await?.into(),
+        })
     }
 }
 
-impl<'a> From<(&'a [u8], &str)> for AttachmentType<'a> {
-    fn from(params: (&'a [u8], &str)) -> AttachmentType<'a> {
+impl From<(Cow<'static, [u8]>, Cow<'static, str>)> for AttachmentType<'static> {
+    fn from((data, filename): (Cow<'static, [u8]>, Cow<'static, str>)) -> Self {
         AttachmentType::Bytes {
-            data: Cow::Borrowed(params.0),
-            filename: params.1.to_string(),
+            data,
+            filename,
         }
     }
 }
```

Comment:
Same as above, I'd recommend replacing as follows:
```rust
impl<'a, S> From<(Cow<'static, [u8]>, S)> for AttachmentType<'a>
where
    S: Into<Cow<'static, str>>,
{
    fn from((data, filename): (Cow<'static, [u8]>, S)) -> Self {
        AttachmentType::Bytes {
            data,
            filename: filename.into(),
        }
    }
}
```
Btw, why did you change it to `AttachmentType<'static>`, was it to coerce `AttachmentType::Bytes(&'static [u8], &'static str)` to be `'static` as well? I'd say with the new impl it's better to loosen the lifetime again.

Reply:
Note that it's currently not possible to change the `Cow<'static, [u8]>` into an `impl Into<Cow>` because the compiler complains about conflicting implementations, citing:
```
upstream crates may add a new impl of trait `From<&tokio::fs::File>` for type `Cow<'static, [u8]>` in future versions
```
Now, I think it's obvious this won't happen, but I don't really like the ergonomics as is, where the user is forced to call `.into()` on their arguments (which normally would be passed inline). I'm in favor of taking a simple `&[u8]` here and collecting into a Vec, but that would mean that `Cow::<'static, u8>::Borrowed` never actually ends up being constructed, so better then to just change the `data` field of the `Bytes` variant to a `Vec<u8>` to avoid confusion.